### PR TITLE
Optimize allocations - shared cursor struct for Put/Get/Bucket

### PR DESCRIPTION
**Pull request by wACKYz**
Each bucket operation creates a new cursor with []itemRef
moving it to the Tx structure allows to share it and avoid allocations.

BenchmarkAlloc1024Op: 1024 operations in one transaction:

```
benchmark                               old ns/op      new ns/op      delta
BenchmarkDBBatchAutomatic-16            3461377        3337566        -3.58%
BenchmarkDBBatchSingle-16               1981311812     1979383688     -0.10%
BenchmarkAlloc1024Op/Get-16             3927441        3772682        -3.94%
BenchmarkAlloc1024Op/PutNew-16          5325346        5190664        -2.53%
BenchmarkAlloc1024Op/PutReplace-16      5140605        5018904        -2.37%
BenchmarkAlloc1024Op/SharedTxGet-16     227576         108000         -52.54%
BenchmarkDBBatchManual10x100-16         21471994       21324567       -0.69%

benchmark                               old allocs     new allocs     delta
BenchmarkDBBatchAutomatic-16            10194          8184           -19.72%
BenchmarkDBBatchSingle-16               56811          54913          -3.34%
BenchmarkAlloc1024Op/Get-16             2110           62             -97.06%
BenchmarkAlloc1024Op/PutNew-16          3377           1328           -60.68%
BenchmarkAlloc1024Op/PutReplace-16      3344           1295           -61.27%
BenchmarkAlloc1024Op/SharedTxGet-16     2048           0              -100.00%
BenchmarkDBBatchManual10x100-16         5937           3981           -32.95%

benchmark                               old bytes     new bytes     delta
BenchmarkDBBatchAutomatic-16            451744        378593        -16.19%
BenchmarkDBBatchSingle-16               28234584      28218672      -0.06%
BenchmarkAlloc1024Op/Get-16             87860         14096         -83.96%
BenchmarkAlloc1024Op/PutNew-16          470349        396572        -15.69%
BenchmarkAlloc1024Op/PutReplace-16      321534        247745        -22.95%
BenchmarkAlloc1024Op/SharedTxGet-16     73728         0             -100.00%
BenchmarkDBBatchManual10x100-16         889315        803483        -9.65%
```

### CODE REVIEW

**Comment by wACKYz** [comment]
Ok as I see it is allowed to use multiple goroutines with a single TX read, I will push fix for it.

**Comment by wACKYz** [comment]
fixed case of multiple goroutines sharing Read TX
benchmarks removed from PR (I can make a separate PR with them if needed)

**Comment by serathius** [comment]
For performance improvements please use [benchstat](https://pkg.go.dev/golang.org/x/perf/cmd/benchstat).
You can learn more about it here https://about.sourcegraph.com/blog/go/gophercon-2019-optimizing-go-code-without-a-blindfold

**Comment by wACKYz** [comment]
```
goos: linux
goarch: amd64
pkg: go.etcd.io/bbolt
cpu: Intel(R) Xeon(R) Gold 6154 CPU @ 3.00GHz
                                  │      org      │              optimized               │
                                  │    sec/op     │    sec/op     vs base                │
DBBatchAutomatic-72                  5.630m ± 14%   5.111m ±  9%   -9.23% (p=0.023 n=10)
DBBatchSingle-72                     242.3m ± 25%   239.6m ± 26%        ~ (p=0.971 n=10)
DB1024OpOnSingleTx/Get-72           2035.1µ ± 10%   825.3µ ± 10%  -59.44% (p=0.000 n=10)
DB1024OpOnSingleTx/PutNew-72         2.385m ± 20%   2.117m ± 43%        ~ (p=0.315 n=10)
DB1024OpOnSingleTx/PutReplace-72     2.142m ± 11%   1.423m ±  7%  -33.59% (p=0.000 n=10)
DB1024OpOnSingleTx/SharedTxGet-72    509.3µ ±  4%   143.4µ ±  2%  -71.84% (p=0.000 n=10)
DBBatchManual10x100-72               5.810m ± 13%   5.548m ± 21%        ~ (p=0.143 n=10)
geomean                              4.575m         3.044m        -33.47%

                                  │     org      │                 optimized                 │
                                  │     B/op     │     B/op      vs base                     │
DBBatchAutomatic-72                 443.0Ki ± 0%   372.1Ki ± 0%   -15.99% (p=0.000 n=10)
DBBatchSingle-72                    17.29Mi ± 5%   17.35Mi ± 5%         ~ (p=0.739 n=10)
DB1024OpOnSingleTx/Get-72           93.27Ki ± 0%   21.10Ki ± 0%   -77.38% (p=0.000 n=10)
DB1024OpOnSingleTx/PutNew-72        467.4Ki ± 0%   395.1Ki ± 0%   -15.47% (p=0.000 n=10)
DB1024OpOnSingleTx/PutReplace-72    321.7Ki ± 0%   249.5Ki ± 0%   -22.46% (p=0.000 n=10)
DB1024OpOnSingleTx/SharedTxGet-72   72.00Ki ± 0%    0.00Ki ± 0%  -100.00% (p=0.000 n=10)
DBBatchManual10x100-72              867.1Ki ± 2%   795.6Ki ± 2%    -8.24% (p=0.000 n=10)
geomean                             490.9Ki                      ?                       ¹ ²
¹ summaries must be >0 to compute geomean
² ratios must be >0 to compute geomean

                                  │     org      │                optimized                 │
                                  │  allocs/op   │  allocs/op   vs base                     │
DBBatchAutomatic-72                 10.194k ± 0%   8.190k ± 0%   -19.66% (p=0.000 n=10)
DBBatchSingle-72                     56.75k ± 0%   54.75k ± 0%    -3.52% (p=0.000 n=10)
DB1024OpOnSingleTx/Get-72           2110.00 ± 0%    62.00 ± 0%   -97.06% (p=0.000 n=10)
DB1024OpOnSingleTx/PutNew-72         3.376k ± 0%   1.327k ± 0%   -60.69% (p=0.000 n=10)
DB1024OpOnSingleTx/PutReplace-72     3.344k ± 0%   1.295k ± 0%   -61.27% (p=0.000 n=10)
DB1024OpOnSingleTx/SharedTxGet-72    2.048k ± 0%   0.000k ± 0%  -100.00% (p=0.000 n=10)
DBBatchManual10x100-72               5.941k ± 1%   3.942k ± 1%   -33.65% (p=0.000 n=10)
geomean                              5.576k                     ?                       ¹ ²
```
The benefit will be visible in large transactions, a single operation/transaction will probably not be noticeable because of  transaction overhead.

Should I include my benchmark in this PR?

**Comment by ptabor** [comment]
Thank you for the PR. The number looks great. At the same time I'm a little bit afraid of putting changes to the critical logic without growing test coverage. We suspect there is already some subtle memory stomping bug and future changes can mask it even further. I would keep it parked / add to an experimental branch if we had such for now.

**Comment by ahrtr** [comment]
Thanks @wACKYz . Sorry for the late response. 

This is the most crucial part of the code. I agree with @ptabor , and it's the reason why I do not review it so far. 

I am also thinking to revert the couple of PRs I merged into master branch for now, and get them included in a temporary branch. WDYT? @ptabor @serathius  We need to focus on adding/improving test, the goal is try to reproduce the data corruption issue.

File: tx.go [hunk 1]
```diff
@@ -587,6 +589,31 @@ func (tx *Tx) Page(id int) (*common.PageInfo, error) {
...
+
+// seekWithSharedCursorThreadSafe should be used in internal functions.
+// Reuses []elemRef stored in Tx to remove unnecessary allocations.
+// It doesn't return a cursor because it's not thread safe
+func (tx *Tx) seekWithSharedCursorThreadSafe(b *Bucket, seek []byte) (key []byte, value []byte, flags uint32) {
+	if atomic.CompareAndSwapUint32(&tx.reusableRefLocked, 0, 1) {
+		// should be 100% compatible with Bucket.Cursor()
+		c := Cursor{bucket: b, stack: tx.reusableRef}
+		key, value, flags = c.seek(seek)
+		tx.reusableRef = c.stack[:0]
```
Comment (cenkalti) [inline tx.go:610]: 4 lines above are duplicated. Can you call `seekWithSharedCursorThreadUnsafe` from here?

**Review by cenkalti — CHANGES_REQUESTED** [review]

**Pull request closed without merging**